### PR TITLE
[Pipeline inference] Modify to tieweight

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
Modify to tieweight

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/pipeline/engine.py [hunk 1]
```diff
@@ -69,9 +69,11 @@ def __init__(
         self.pp_size = pp_size
         self.pg_mesh = ProcessGroupMesh(pp_size)
         self.stage_manager = PipelineStageManager(self.pg_mesh, 0, True)
-        self.mb_manager = MicroBatchManager(new_length, micro_batch_size, micro_batch_buffer_size or pp_size)
+        self.mb_manager = MicroBatchManager(self.stage_manager.stage, new_length, micro_batch_size,
+                                            micro_batch_buffer_size or pp_size)
         self.schedule = GenerateSchedule(self.stage_manager, self.mb_manager)
         self.model = pp_model or self._shardformer(model, model_policy)
+        print(dist.get_rank(), torch.cuda.memory_allocated())
```
Comment (ver217) [inline colossalai/inference/pipeline/engine.py:76]: Please remove this

File: colossalai/inference/pipeline/microbatch_manager.py [hunk 2]
```diff
@@ -16,118 +16,171 @@ class MicroBatchDescription():
...
         self.kv_cache = kv_cache
 
+    @property
+    def state(self):
+        """
+        Return the state of current micro batch, when current length is equal to target length,
+        the state is DONE, otherwise GENERATE
+
+        """
+        if self.cur_length == self.target_length:
```
Comment (ver217) [inline colossalai/inference/pipeline/microbatch_manager.py:43]: This still doesn't support early stopping, right? If so, please add TODO here.

File: colossalai/inference/pipeline/microbatch_manager.py [hunk 3]
```diff
@@ -16,118 +16,171 @@ class MicroBatchDescription():
...
-        otherwise the sequence length is `kv_cache[0][0].shape[-2]` plus 1
+        Return the current sequnence length of micro batch
 
         """
-        if len(self.kv_cache) == 0:
+        pass
+
+
+class HeadMicroBatchDescription(MicroBatchDescription):
+
```
Comment (ver217) [inline colossalai/inference/pipeline/microbatch_manager.py:58]: What's the difference of three types of batch description?

File: colossalai/inference/pipeline/microbatch_manager.py [hunk 4]
```diff
@@ -16,118 +16,171 @@ class MicroBatchDescription():
...
-        self.next()
-        return state
+            self._add_descrption(inputs_dict, output_dict)
+        self.cur_descrption.update(output_dict, new_token)
+        return self.cur_state
 
-    def next(self):
-        self.idx = (self.idx + 1) % self.buffer_size
+    def export_new_tokens(self):
+        list = [i.new_tokens.tolist()[0] for i in self.mb_descrption_buffer.values()]
```
Comment (ver217) [inline colossalai/inference/pipeline/microbatch_manager.py:159]: `list` is a reserved keyword. Please DON'T use it.

Comment (ver217) [reply colossalai/inference/pipeline/microbatch_manager.py:159]: By the way, `tolist()[0]` is not memory efficient. Can you slice then tolist?

File: colossalai/inference/pipeline/engine.py (hunk 1 above)
Comment (FoolPlayer) [reply colossalai/inference/pipeline/engine.py:76]: update

File: colossalai/inference/pipeline/microbatch_manager.py (hunk 2 above)
Comment (FoolPlayer) [reply colossalai/inference/pipeline/microbatch_manager.py:43]: update

File: colossalai/inference/pipeline/microbatch_manager.py (hunk 3 above)
Comment (FoolPlayer) [reply colossalai/inference/pipeline/microbatch_manager.py:58]: `HeadMicroBatchDescription` is used to describe the infomation of the first stage, and `BodyMicroBatchDescription` is used for other stages, `HeadMicroBatchDescription` has `token` and `mask` attribute, and also the update method is different. I make them separate to make the code and logistic more clear.

File: colossalai/inference/pipeline/microbatch_manager.py (hunk 4 above)
Comment (FoolPlayer) [reply colossalai/inference/pipeline/microbatch_manager.py:159]: update

File: colossalai/inference/pipeline/microbatch_manager.py (hunk 3 above)
Comment (ver217) [reply colossalai/inference/pipeline/microbatch_manager.py:59]: Can you add doctring to illustrate this?

**Review by ver217 — APPROVED** [review]

**Pull request merged**